### ISSUE TITLE
HTTP/2 StreamByteDistributors allowing headers to be written when no data can be written

### ISSUE THREAD

**Issue by Scottmitch** [netty/netty#4585]
When the discussing the design of the UniformStreamByteDistributor (https://github.com/netty/netty/issues/4489#issuecomment-158472088) the intention was to only send the initial headers frame if we could also send data for that stream. The rational being that this would lead to better data locality.  However the current implementation may not achieve this under all circumstances. 

[UniformStreamByteDistributor checks StreamState.streamableBytes > 0](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java#L100) to enforce this condition. Streamable bytes could be 0 because the window is collapsed, and not just because there are zero pending flow controlled bytes to send (streamable bytes = `max(0, min(pendingBytes, window));`). So it is possible that we let a write slip because the window has collapsed and the stream has non zero length flow controlled frames to send (example stream frame queue = `[Headers Frames(0 size), Data Frame(1Mb size)]`, `window=0`, `pending=1Mb`).

If we still think this is a good approach we should fix this implementation and apply the same constraints to `WeightedFairQueueByteDistributor`.

**Comment by Scottmitch** [comment]
@nmittler - FYI

**Comment by nmittler** [comment]
@Scottmitch agreed! This is lower priority, but should still probably be fixed.

It seems that we should give the distributor the raw data (stream window, numBytes, hasFrame) so it can make a better decision. WDYT?

**Comment by Scottmitch** [comment]
> It seems that we should give the distributor the raw data 

@nmittler - +1. This is what I ended up doing.

**Comment by normanmaurer** [comment]
@Scottmitch @nmittler was this fixed ?

**Comment by Scottmitch** [comment]
I don't think so ... still lower priority though.

**Comment by durigon** [comment]
If we do not allow writing long-length objects(streamableBytes ... ) in HTTP/2 flow control(#4283), 
it may infinite loop in UniformStreamByteDistributor([UniformStreamByteDistributor checks StreamState.streamableBytes > 0](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java#L100)).

In the current version, flow controller is required so that HTTP/2 stream pending bytes do not exceed the integer range.

WDYT?
